Add support for retrieving logger level number to solve #36

jsumners opened the pull request:
Once more as if I know what I'm doing.

Also in this commit, it looks like the Readme had some trailing spaces. My editor trimmed those. I didn't notice until I looked at this pull request.

davidmarkclements commented:
how come this is this a diff PR to #37 ?

davidmarkclements commented:
also - any objections to shortening to `levelInt` - since int is commonly understood abbr. of interger? / cc @mcollina

jsumners commented:
My merge of upstream resulted in a mess. So I scrapped the whole thing and recloned. I thought it would have deleted the other pull request since the repo no longer exists. 

Personally, I find 'levelInt' hard to read due to 'l' and 'I' being so similar in a lot of fonts. At least the full word makes it clearer.

davidmarkclements commented:
ok cool - true - but similarly applies to levelInteger even given the extra context - any takers for `levelNum` or `nLevel`? 

In terms of the implementation it's more than required - we already have a `_level` property that stores the numerical level - you could simply rename that to `levelInteger` (or `levelNum`) and then there's zero-cost to the API addition

jsumners commented:
So you would rather make `_level` a public property? Just let me know what you want it named. I'll wait for a consensus before submitting any more changes.

davidmarkclements commented:
I think so - I'll defer to @mcollina for consensus

mcollina commented:
I will go for nLevel.

satazor commented:
I prefer  `levelNum`

mcollina commented:
levelNum is ok as well, let's go for it. I don't really care as long as the
docs is complete.

We should also expose the mappings between numbers and strings.
Il giorno lun 2 mag 2016 alle 20:52 André Cruz <notifications@github.com>
ha scritto:

> I prefer levelNum
>
> —
> You are receiving this because you were mentioned.
>
>
> Reply to this email directly or view it on GitHub
> <https://github.com/mcollina/pino/pull/39#issuecomment-216326264>
>

jsumners commented:
Here's what I did:

+ `levelNum` -- returns current logger level number
+ `levelNumbers` -- returns `levels` hash, `string => number`
+ `levelStrings` -- returns `nums` hash, `number => string`

You may want to run the benchmarks. On my system, which is doing way too much right now, these changes incurred a slight penalty (~5ms).

davidmarkclements commented:
The cost is going to compound for child loggers, instead we should defineProperty on the prototype with a getter that returns (a copy of) the object

I suggest we just have a `levels` property with a getter that returns an object:

```
get levels => {strings, numbers}
```

Perhaps we should call the, values and labels instead though?

jsumners commented:
Will do.

On Monday, May 2, 2016, David Mark Clements <notifications@github.com>
wrote:

> The cost is going to compound for child loggers, instead we should
> defineProperty on the prototype with a getter that returns (a copy of) the
> object
>
> I suggest we just have a levels property with a getter that returns an
> object:
>
> get levels => {strings, numbers}
>
> Perhaps we should call the, values and labels instead though?
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly or view it on GitHub
> <https://github.com/mcollina/pino/pull/39#issuecomment-216358231>
>


-- 
James Sumners
http://james.sumners.info/ (technical profile)
http://jrfom.com/ (personal site)
http://haplo.bandcamp.com/ (band page)

mcollina commented:
I think `levelNumbers` and `levelStrings` should be properties of `pino`, not of the logger. That should not introduce any penalty.

jsumners commented:
Okay, I moved the mappings to `pino.levels.numbers` and `pino.levels.strings`. I also updated `Pino.levelNum` to set the level if a number is assigned to it, and made `Pino._setLevel` use `Pino.levelNum`. Defining `Pino.levelNum` as underscore function and then defining them as a property, as is done with `level`, resulted in a significant hit to performance.  As you can see from my runs of `benchmarks/basic`, the way I have done it results in a negligible difference (statistically irrelevant).

My commit:
```
benchBunyan*10000: 2121ms
benchWinston*10000: 3507ms
benchBole*10000: 3425ms
benchPino*10000: 503ms
benchPinoExreme*10000: 169ms
benchBunyan*10000: 2083ms
benchWinston*10000: 3366ms
benchBole*10000: 3318ms
benchPino*10000: 490ms
benchPinoExreme*10000: 154ms
```

Upstream commit (041b116):
```
benchBunyan*10000: 2034ms
benchWinston*10000: 3450ms
benchBole*10000: 3184ms
benchPino*10000: 502ms
benchPinoExreme*10000: 162ms
benchBunyan*10000: 2027ms
benchWinston*10000: 3409ms
benchBole*10000: 3153ms
benchPino*10000: 502ms
benchPinoExreme*10000: 166ms
```

davidmarkclements commented:
@jsumners great work man
@mcollina & @jsumners: are we happy with the terminology here: strings, numbers, levelNum? As opposed to labels, values, levelVal

mcollina commented:
After a good night of sleep, the labels and values terminology sound much better. @jsumners can you update this (sorry for all the changes, consensus takes time).

mcollina commented on README.md:
````diff
@@ -331,10 +334,35 @@ object, all its properties will be included in the JSON line.
+```
+
+<a name="levelStrings"></a>
+### logger.levels.strings
````
this should be `'pino'`, right?

jsumners replied:
Oops. I fixed it during the re-label.

davidmarkclements commented on pino.js:
```diff
@@ -91,26 +100,35 @@ Pino.prototype.info = genLog(levels.info)
+      if (num > levels[key]) {
+        this[key] = noop
+        continue
+      }
```
just noticed you remove the check against whether the key has been set to no-op

is there a perf difference to removing this check? we might have to verify perf difference by changing the levels to trigger the prop overwriting

my thought at the time was it's better not to shadow the prototype if possible (in view of the way hidden classes work)

jsumners replied:
Yes, the extra check was adding at least 5ms.

jsumners commented:
Labeling has been changed to val, values, labels.

davidmarkclements commented:
lol@commit

mcollina commented:
That's awesome!!  Thank so much for the good work here.

It's LGTM for me.

@davidmarkclements?

davidmarkclements commented:
@mcollina yep - although maybe we should squash the commits here into one for brevity - @jsumners are you happy to squash commits into single?

jsumners commented:
@davidmarkclements I do not know how to do that

davidmarkclements commented:
it's basically an interactive rebase, change all the picks apart from one to squash: 

http://gitready.com/advanced/2009/02/10/squashing-commits-with-rebase.html

davidmarkclements commented:
@jsumners - nice one man - I'll just wait for ci builds then merge - thanks for all your effort, really appreciated!

jsumners commented:
@davidmarkclements cool. I hope it's all correct. That's the first time I've successfully done that rebase business (and this took several tries). I'm a "keep all history as it was written" person.

mcollina commented:
I'm glad you learned something new @jsumners!
BTW, that's (part of) the process we use for contributing to node core :).

davidmarkclements commented:
I think in this case there was a lot of vacillation adding noise imo - usually wouldn't be overly strict with a couple of commits per pr

* merged
* 2.3.0 published
* v2.3.0 gh release done

mcollina commented:
🎉 🎉 !
